commit fa039d5f6b126fbd65eefa05db2f67e44df8f121
Author: Timo Warns <Warns@pre-sense.de>
Date:   Fri May 6 13:47:35 2011 +0200

    Validate size of EFI GUID partition entries.
    
    Otherwise corrupted EFI partition tables can cause total confusion.
    
    Signed-off-by: Timo Warns <warns@pre-sense.de>
    Cc: stable@kernel.org
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
    [dannf: backported to Debian's 2.6.32]

diff -urpN linux-source-2.6.32.orig/fs/partitions/efi.c linux-source-2.6.32/fs/partitions/efi.c
--- linux-source-2.6.32.orig/fs/partitions/efi.c	2011-05-03 09:28:49.000000000 -0600
+++ linux-source-2.6.32/fs/partitions/efi.c	2011-05-17 00:36:47.701635525 -0600
@@ -349,6 +349,12 @@ is_gpt_valid(struct block_device *bdev,
 		goto fail;
 	}
 
+	/* Check that sizeof_partition_entry has the correct value */
+	if (le32_to_cpu((*gpt)->sizeof_partition_entry) != sizeof(gpt_entry)) {
+		pr_debug("GUID Partitition Entry Size check failed.\n");
+		goto fail;
+	}
+
 	if (!(*ptes = alloc_read_gpt_entries(bdev, *gpt)))
 		goto fail;
 
